torch/csrc/distributed/autograd/engine/dist_engine.cpp

```diff
@@ -254,17 +248,71 @@ void DistEngine::computeDependencies(
   autogradContext->setGraphTask(std::move(graphTask));
 }
 
+std::shared_ptr<FutureVariableList> DistEngine::execute_graph_task_until_ready_queue_empty(
+    const std::shared_ptr<GraphTask>& graph_task,
+    std::shared_ptr<Node> root_to_execute,
+    bool incrementOutstandingTasks) {
+  engine_.initialize_device_threads_pool();
+  // Create a ready queue per call to traverse the graph_task from root_to_execute
+  // This allow concurrent execution of the same GraphTask from different threads
+  std::shared_ptr<ReadyQueue> cpu_ready_queue = std::make_shared<ReadyQueue>();
+  cpu_ready_queue->push(
+      NodeTask(graph_task, std::move(root_to_execute), InputBuffer(0)),
+      incrementOutstandingTasks);
+
+  torch::autograd::set_device(torch::autograd::CPU_DEVICE);
+  graph_task->owner_ = torch::autograd::CPU_DEVICE;
+  while(!cpu_ready_queue->empty()) {
+    std::shared_ptr<GraphTask> local_graph_task;
+    {
+      // Scope this block of execution since NodeTask is not needed after this
+      // block and can be deallocated (release any references to grad tensors
+      // as part of inputs_)
+      NodeTask task = cpu_ready_queue->pop();
+      if (!(local_graph_task = task.base_.lock())) {
+        continue;
+      }
+      if (task.fn_ && !local_graph_task->has_error_.load()) {
+        AutoGradMode grad_mode(local_graph_task->grad_mode_);
+        try {
+          engine_.evaluate_function(local_graph_task, task.fn_.get(), task.inputs_, cpu_ready_queue);
```

Comment:
Not sure if you this is supposed to support reentrant autograd. If so, I think it is missing an `init_thread_local_ready_queue()`. Otherwise, it's fine.

Reply:
I think in the dist engine call phase, it's a bit hard now to support reentrant call from `distributed::backward()` level, but for local reentrant backward that dist backward calls, it might be possible to support it. The tricky here is, we are using a locally created ready_queue per `execute_until_ready_queue_empty`, to execute the same graph task concurrently, if we `init_thread_local_ready_queue()`, there's chance that the thread local ready_queue get overwrited by two different threads concurrently (e.g. thread A and B doing local reentrant backward concurrently). We might need to think more on how to handle local reentrant backward calls indeed, right now the approach is to use a new ready queue whenever local reentrant calls `Engine::execute`. I would leave it as a follow up task.

Reply:
But this is a thread local variable. so different thread should not overwrite each other right?

I am fine if this is not supported, just wanted to point it out. Maybe you want to add a comment to make sure this is clear.

Reply:
Oh actually you are right, since it's a thread local it will be preserved correctly. Let me try to add it :)